src/mds/Locker.cc

```diff
@@ -5536,6 +5546,7 @@ void Locker::local_xlock_finish(const MutationImpl::lock_iterator& it, MutationI
   mut->locks.erase(it);
 
   lock->finish_waiters(SimpleLock::WAIT_STABLE |
```

Comment:
Should be `XLOCK | WR` only

Reply:
Optimization: if there is another `XLOCK` waiter we should only finish that. Finishing `WR` waiters in the presence of an `XLOCK` waiter is a waste of time, as they won't be able to wrlock anyway.

Reply:
```suggestion
  lock->finish_waiters(SimpleLock::WAIT_XLOCK | (lock->can_wrlock() ? SimpleLock::WAIT_WR | 0));
```

Reply:
With the new definition of `can_wrlock()` `w1` and `w2` will wait again because `x1` is among the waiters

Reply:
OK, sorry, I had to think twice for this. You are right, there will be no xlocker on the list for w1 and w2 because it will be taken.

Reply:
Either way, `XLOCK | WR` should suffice.